### ISSUE TITLE
Data::Dumper silently fails, exiting the script

### ISSUE THREAD

**Issue by j-mckim** [Perl/perl5#18614]
When dumping a structure that contains certain Unicode characters, Data::Dumper silently exits the script. The remainder of the structure is not dumped. The rest of the script is not executed.

More seriously, eval{} doesn't intercept the exit.

There's an interaction between Data::Dumper and binmode(). If the latter isn't specified, Dumper works. If it's present, Dumper fails as described.

Versions are: perl 5 version 30, and Data::Dumper version 2.174.

Attached is a short script that triggers the bug.

[bug-data-dumper.pl.txt](https://github.com/Perl/perl5/files/6069516/bug-data-dumper.pl.txt)

**Comment by jkeenan** [comment]
Confirmed; present from at least perl-5.8.4.

**Comment by hvds** [comment]
It is not reproducing for me here.

Do you get the same effect if you replace the literal Unicode character with the corresponding `\x{41f}` escape?

What is the exit code of the script when it fails? (On a Unix system you can check that with `echo $?` directly after the script completes; 0 means it (thinks it) completed ok; 255 usually means a Perl-level crash; any other value usually means a C-level crash.)

Can you catch the failure with a `__DIE__` handler, to get a Perl-level stacktrace? Insert at the start:
```
  use Carp;
  $SIG{__DIE__} = sub { Carp::confess(@_) };
```

Can you run it in a C-level debugger? If it dies with a signal such as SEGV, the debugger should stop at the point the signal is raised;  if so you may be able to get a C-level stacktrace:
```
% gdb -q --args /opt/v5.30.0-d/bin/perl ~/bug-data-dumper.pl.txt
Reading symbols from /opt/v5.30.0-d/bin/perl...done.
(gdb) run
Starting program: /opt/v5.30.0-d/bin/perl /home/hv/bug-data-dumper.pl.txt
$VAR1 = [
          'some stuff',
          " \x{41f}  non-regex with problem char dumps correctly",
          qr/ è® regex with other unicode char dumps correctly /u,
          qr/ Ð regex with problem char; this part is not dumped /u,
          'yet more stuff not dumped'
        ];
Hello! I didn't die!
[Inferior 1 (process 6003) exited normally]
(gdb) where
No stack.
(gdb) exit
% 
```

(This is most useful using a perl built with debugging, but even without that it would give us a clue.)

**Comment by arc** [comment]
I can't reproduce the silent exit:

```
$ perl bug-data-dumper.pl.txt
$VAR1 = [
          'some stuff',
          " \x{41f}  non-regex with problem char dumps correctly",
          qr/ è® regex with other unicode char dumps correctly /u,
          qr/ Ð regex with problem char; this part is not dumped /u,
          'yet more stuff not dumped'
        ];
Hello! I didn't die!
$ perl bug-data-dumper.pl.txt | hexdump -C
00000000  24 56 41 52 31 20 3d 20  5b 0a 20 20 20 20 20 20  |$VAR1 = [.      |
00000010  20 20 20 20 27 73 6f 6d  65 20 73 74 75 66 66 27  |    'some stuff'|
00000020  2c 0a 20 20 20 20 20 20  20 20 20 20 22 20 5c 78  |,.          " \x|
00000030  7b 34 31 66 7d 20 20 6e  6f 6e 2d 72 65 67 65 78  |{41f}  non-regex|
00000040  20 77 69 74 68 20 70 72  6f 62 6c 65 6d 20 63 68  | with problem ch|
00000050  61 72 20 64 75 6d 70 73  20 63 6f 72 72 65 63 74  |ar dumps correct|
00000060  6c 79 22 2c 0a 20 20 20  20 20 20 20 20 20 20 71  |ly",.          q|
00000070  72 2f 20 c3 a8 c2 ae c2  80 20 72 65 67 65 78 20  |r/ ...... regex |
00000080  77 69 74 68 20 6f 74 68  65 72 20 75 6e 69 63 6f  |with other unico|
00000090  64 65 20 63 68 61 72 20  64 75 6d 70 73 20 63 6f  |de char dumps co|
000000a0  72 72 65 63 74 6c 79 20  2f 75 2c 0a 20 20 20 20  |rrectly /u,.    |
000000b0  20 20 20 20 20 20 71 72  2f 20 c3 90 c2 9f 20 72  |      qr/ .... r|
000000c0  65 67 65 78 20 77 69 74  68 20 70 72 6f 62 6c 65  |egex with proble|
000000d0  6d 20 63 68 61 72 3b 20  74 68 69 73 20 70 61 72  |m char; this par|
000000e0  74 20 69 73 20 6e 6f 74  20 64 75 6d 70 65 64 20  |t is not dumped |
000000f0  2f 75 2c 0a 20 20 20 20  20 20 20 20 20 20 27 79  |/u,.          'y|
00000100  65 74 20 6d 6f 72 65 20  73 74 75 66 66 20 6e 6f  |et more stuff no|
00000110  74 20 64 75 6d 70 65 64  27 0a 20 20 20 20 20 20  |t dumped'.      |
00000120  20 20 5d 3b 0a 48 65 6c  6c 6f 21 20 49 20 64 69  |  ];.Hello! I di|
00000130  64 6e 27 74 20 64 69 65  21 0a                    |dn't die!.|
0000013a
```

However, the output is clearly garbled, and I think I can see why. Pull request incoming.

**Comment by j-mckim** [comment]
More data: It was not a silent exit via Data::Dumper. Instead, some
characters were output that put my terminal into a silent mode, making it
cease to display anything more. At least until the normal exit to the shell
prompt.

I was using gnome-terminal. I just tried it on a MacOS terminal and the bug
was not triggered.

FWIW, this is a demo of the sequence, outside the context of Data::Dumper.
It triggers on the gnome terminal, and not on the MacOS terminal.

print "hello \xc3\x90\xc2\x9f abcd\nxyzzy\nplugh\n";

**Comment by hvds** [comment]
> More data: It was not a silent exit via Data::Dumper. Instead, some characters were output that put my terminal into a silent mode, making it cease to display anything more.

Ah cool, in that case it is likely @arc's patch will fix it for you; hopefully @xsawyerx or a delegate will roll it out to CPAN once it has had some smoke testing, so you won't need to wait for a new release of perl.

**Comment by jkeenan** [comment]
> I can't reproduce the silent exit:
> 
> ```
> $ perl bug-data-dumper.pl.txt
> $VAR1 = [
>           'some stuff',
>           " \x{41f}  non-regex with problem char dumps correctly",
>           qr/ è®� regex with other unicode char dumps correctly /u,
>           qr/ Ð� regex with problem char; this part is not dumped /u,
>           'yet more stuff not dumped'
>         ];
> Hello! I didn't die!
> $ perl bug-data-dumper.pl.txt | hexdump -C
> 00000000  24 56 41 52 31 20 3d 20  5b 0a 20 20 20 20 20 20  |$VAR1 = [.      |
> 00000010  20 20 20 20 27 73 6f 6d  65 20 73 74 75 66 66 27  |    'some stuff'|
> 00000020  2c 0a 20 20 20 20 20 20  20 20 20 20 22 20 5c 78  |,.          " \x|
> 00000030  7b 34 31 66 7d 20 20 6e  6f 6e 2d 72 65 67 65 78  |{41f}  non-regex|
> 00000040  20 77 69 74 68 20 70 72  6f 62 6c 65 6d 20 63 68  | with problem ch|
> 00000050  61 72 20 64 75 6d 70 73  20 63 6f 72 72 65 63 74  |ar dumps correct|
> 00000060  6c 79 22 2c 0a 20 20 20  20 20 20 20 20 20 20 71  |ly",.          q|
> 00000070  72 2f 20 c3 a8 c2 ae c2  80 20 72 65 67 65 78 20  |r/ ...... regex |
> 00000080  77 69 74 68 20 6f 74 68  65 72 20 75 6e 69 63 6f  |with other unico|
> 00000090  64 65 20 63 68 61 72 20  64 75 6d 70 73 20 63 6f  |de char dumps co|
> 000000a0  72 72 65 63 74 6c 79 20  2f 75 2c 0a 20 20 20 20  |rrectly /u,.    |
> 000000b0  20 20 20 20 20 20 71 72  2f 20 c3 90 c2 9f 20 72  |      qr/ .... r|
> 000000c0  65 67 65 78 20 77 69 74  68 20 70 72 6f 62 6c 65  |egex with proble|
> 000000d0  6d 20 63 68 61 72 3b 20  74 68 69 73 20 70 61 72  |m char; this par|
> 000000e0  74 20 69 73 20 6e 6f 74  20 64 75 6d 70 65 64 20  |t is not dumped |
> 000000f0  2f 75 2c 0a 20 20 20 20  20 20 20 20 20 20 27 79  |/u,.          'y|
> 00000100  65 74 20 6d 6f 72 65 20  73 74 75 66 66 20 6e 6f  |et more stuff no|
> 00000110  74 20 64 75 6d 70 65 64  27 0a 20 20 20 20 20 20  |t dumped'.      |
> 00000120  20 20 5d 3b 0a 48 65 6c  6c 6f 21 20 49 20 64 69  |  ];.Hello! I di|
> 00000130  64 6e 27 74 20 64 69 65  21 0a                    |dn't die!.|
> 0000013a
> ```
> 
> However, the output is clearly garbled, and I think I can see why. Pull request incoming.

@arc do you have a timeline for this pull request?

Thank you very much.
Jim Keenan

**Comment by arc** [comment]
The PR was #18619, which was merged as commit 33bc847050ad68bb79f1e04db9100e25017348e1 — my apologies for not closing this ticket at the time.
